### ISSUE TITLE
[Bug]: TS type aliases can't be augmented

### ISSUE THREAD

**Issue by fabis94** [storybookjs/storybook#19950]
### Describe the bug

AFAIK TypeScript type aliases (e.g. `type Story = ...`) can't be [augmented](https://www.typescriptlang.org/docs/handbook/declaration-merging.html), as opposed to TypeScript interfaces. This is a pretty big problem when it comes to Storybook types, because it appears that pretty much all exported types are done using `type` aliases.

What this means in practice is that you can't extend the context or any other type with your own properties, and they will always be typed as `any` unless if you force cast them.

### To Reproduce

```shell
Pretty much all (with minor exceptions) of the types seem to be defined with `type` aliases instead of `interfaces` so just check any TS file in the repo. As an example: https://github.com/storybookjs/storybook/blob/next/code/renderers/server/src/types.ts
```

### System

```shell
System:
    OS: Linux 5.15 Ubuntu 20.04.2 LTS (Focal Fossa)
    CPU: (6) x64 AMD Ryzen 9 5900X 12-Core Processor
  Binaries:
    Node: 16.14.0 - ~/.nvm/versions/node/v16.14.0/bin/node
    Yarn: 3.2.0 - ~/.nvm/versions/node/v16.14.0/bin/yarn
    npm: 8.3.1 - ~/.nvm/versions/node/v16.14.0/bin/npm
  Browsers:
    Chrome: 100.0.4896.75
  npmPackages:
    @storybook/addon-a11y: ^7.0.0-alpha.42 => 7.0.0-alpha.43 
    @storybook/addon-docs: ^7.0.0-alpha.42 => 7.0.0-alpha.42 
    @storybook/addon-essentials: ^7.0.0-alpha.42 => 7.0.0-alpha.42 
    @storybook/addon-interactions: ^7.0.0-alpha.42 => 7.0.0-alpha.42 
    @storybook/addon-links: ^7.0.0-alpha.42 => 7.0.0-alpha.42 
    @storybook/jest: ^0.0.10 => 0.0.10 
    @storybook/test-runner: 0.10.0-next.1 => 0.10.0-next.1 
    @storybook/testing-library: ^0.0.13 => 0.0.13 
    @storybook/vue3: ^7.0.0-alpha.42 => 7.0.0-alpha.42 
    @storybook/vue3-vite: ^7.0.0-alpha.42 => 7.0.0-alpha.42
```

**Comment by valentinpalkovic** [comment]
@kasperpeulen 👆

**Comment by kasperpeulen** [comment]
@fabis94 In the latest alpha, `Parameters` is an interface, which you can augment now:

```ts
declare module '@storybook/react' {
  interface Parameters extends A, B, C, D {}
}
```

Or there any other interfaces that we export that you would like to augment? @fabis94

**Comment by fabis94** [comment]
@kasperpeulen `StoryContext` is the one that prompted me to create this issue - you can usually add stuff to the context through decorators.

Also make the `VueFramework` type that is commonly used as a generic argument for various Storybook types exported out from the package, because otherwise it's very difficult to use some of the generic Storybook storybook types to create new ones

**Comment by kasperpeulen** [comment]
Ah I see. I think it is in general good if we make StoryContext an interface. But I'm unsure if we encourage people to add arbitrary properties to the context.

Looking at this comment:

https://github.com/ComponentDriven/csf/blob/b170e3d459d5619384fdf181b91c0867c62fbd7a/src/story.ts#L109-L115
```ts
export type StoryContextUpdate<TArgs = Args> = {
  args?: TArgs;
  globals?: Globals;
  // NOTE: it is currently possibly to add *any* key you like to the context
  // (although you cannot override the basic keys). This will likely be removed in future.
  [key: string]: any;
};
```

What do you guys think @tmeasday @yannbf? Maybe we should recommend using `parameters` for this?

**Comment by fabis94** [comment]
This is the reason why I had to change the context, I don't think you can achieve the same with just parameters: https://craigbaldwin.com/blog/updating-args-storybook-vue/

**Comment by tmeasday** [comment]
> Maybe we should recommend using parameters for this?

I don't think so, parameters are the one thing you cannot update in a decorator[1] -- the idea of parameters being that they are static and available everywhere (i.e on the manager).

I think args are the ideal place for this callback to go. The pattern we'd recommend for that article is pretty close @fabis94 but I think a bit simpler (I'm no Vue expert so apologies if this is garbled) -- basically, why have the extra level of indirection of passing the `updateArgs` function around, instead the decorator can just do:

```js
(story, context) => {
  const [_, updateArgs] = useArgs()
  const handleQuantityUpdate = (quantity) => updateArgs({ quantity });
  return story({ args: {  handleQuantityUpdate }  });
},
```

Note FTR that you don't need to pass the entire context to `storyFn` or the entire args to `updateArgs`.

[1] Although really you shouldn't be able to update some other stuff like globals too I think.

**Comment by fabis94** [comment]
Ok then it would be nice to be able to augment the Args type as well. Also would be nice to make it possible to modify its contents through a generic argument so that in stories you get type checking when filling out Story.args
